linux/ibus-keyman/src/engine.c

```diff
@@ -595,25 +596,24 @@ ibus_keyman_engine_process_key_event (IBusEngine     *engine,
 
     g_message("DAR: ibus_keyman_engine_process_key_event - keyval=%02i, keycode=%02i, state=%02x", keyval, keycode, state);
 
-    if (state & IBUS_RELEASE_MASK)
-    {
-        switch(keycode) {
-            case KEYMAN_LCTRL:
-                keyman->lctrl_pressed = FALSE;
-                break;
-            case KEYMAN_RCTRL:
-                keyman->rctrl_pressed = FALSE;
-                break;
-            case KEYMAN_LALT:
-                keyman->lalt_pressed = FALSE;
-                break;
-            case KEYMAN_RALT:
-                keyman->ralt_pressed = FALSE;
-                break;
-            default:
-                break;
-        }
-        return FALSE;
+    gboolean isKeyDown = TRUE;
+
+    if (state & IBUS_RELEASE_MASK) {
+      isKeyDown = FALSE;
+      switch (keycode) {
+      case KEYMAN_LCTRL:
+        keyman->lctrl_pressed = FALSE;
+        break;
+      case KEYMAN_RCTRL:
+        keyman->rctrl_pressed = FALSE;
+        break;
+      case KEYMAN_LALT:
+        keyman->lalt_pressed = FALSE;
+        break;
+      case KEYMAN_RALT:
+        keyman->ralt_pressed = FALSE;
+        break;
+      }
     }
 
     if (keycode < 0 || keycode > 255)
```

Comment:
I note that we test `keycode_to_vk[keycode] == 0` on line 625 and that KEYMAN_LCTRL, KEYMAN_RCTRL, KEYMAN_LALT and KEYMAN_RALT are tested within that. This means that we'll never reset the flags as we no longer `return FALSE` above.

We should refactor the switch, outside both of the `if()` blocks:

```c++
      switch (keycode) {
      case KEYMAN_LCTRL:
        keyman->lctrl_pressed = isKeyDown;
        break;
      case KEYMAN_RCTRL:
        keyman->rctrl_pressed = isKeyDown;
        break;
      case KEYMAN_LALT:
        keyman->lalt_pressed = isKeyDown;
        break;
      case KEYMAN_RALT:
        keyman->ralt_pressed = isKeyDown;
        break;
      }
```

And then instead of 

```c++
    gboolean isKeyDown = TRUE;

    if (state & IBUS_RELEASE_MASK) {
      isKeyDown = FALSE;
    }
```

we can do

```c++
    gboolean isKeyDown = !!(state & IBUS_RELEASE_MASK);
```

I also wonder why those four specific shift codes are not given a vk code in the `keycode_to_vk` map, because we have `VK_LCONTROL`, `VK_RCONTROL`, `VK_LMENU`, `VK_RMENU` which correspond. But that may be something to explore in a separate PR?

Reply:
done. Yes, I as well was wondering why we don't handle those four keys. But better to do in a separate PR.